Fix Get-WinEvent -FilterHashtable to work with named event data field

iSazonov opened the pull request:
Now FilterHashtable parameter in Get-WinEvent for named data field:
1. generates a valid query, ex. @{Logname="System";Param="a"}
2. generates a valid query for multiple values, ex.
@{Logname="System";Param="a","b"}

Now I need help in creating tests #2368

Close #2327

iSazonov committed:
Fix Get-WinEvent -FilterHashtable to work with named event data field

Now FilterHashtable parameter in Get-WinEvent for named data field:
1. generates a valid query, ex. @{Logname="System";Param="a"}
2. generates a valid query for multiple values, ex.
@{Logname="System";Param="a","b"}

msftclas commented:
Hi __@iSazonov__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

PaulHigin commented on src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs:
```diff
@@ -1538,6 +1542,41 @@ private string HandleDataHashValue(Object value)
+        private string HandleNamedDataHashValue(String key, Object value)
+        {
+            string ret = "";
+            if (value is Array)
```
This is doing two casts, once with the is keyword and again with the explicit cast.  We normally use this pattern to avoid the second cast:
Array dataArray = value as Array;
if (dataArray != null)
{
...

iSazonov replied:
Paul thanks for code review!
I agree but the entire file uses the old technique!!! I suppose that it makes no sense to change the code in a single place, and there should be a separate issue and PR.

PaulHigin replied:
Good point.  I didn't realize this pattern was used in other places in the code, but I now see it is used in five other places.  This must be old code because it is not a good or common pattern that we use.  I feel it is worth changing in these five places.  Can you make this part of your change?  At the very least I feel that your new change should use the common pattern even if we create a new issue to clean up the rest of the code.

daxian-dbw replied:
> At the very least I feel that your new change should use the common pattern even if we create a new issue to clean up the rest of the code.

I agree with @PaulHigin. Using the recommended pattern in your new code can serve as a good example for anyone who'd come later to fix the rest of code.

iSazonov replied:
Fixed!

daxian-dbw replied:
Thanks!

PaulHigin commented on src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs:
```diff
@@ -1538,6 +1542,41 @@ private string HandleDataHashValue(Object value)
+            if (value is Array)
+            {
+                Array dataArray = (Array)(value);
+                ret += "(";
```
Using StringBuilder to create the string would be more efficient here.

iSazonov replied:
Again this is for new Issue and RP I suppose.

iSazonov replied:
Again this is for new Issue and RP I suppose.

daxian-dbw replied:
Would you like to address this comment as well (at least in your newly added method `HandleNamedDataHashValue`)?

iSazonov replied:
@daxian-dbw I oppose for two reasons:
1. As mentioned, the code is very old and needed to be revised entirely. I don't like the idea of disguising one hole out of thirty.
2. Soon I plan to expand the cmdlet and add support Suppress. To do this, I need to completely rewrite the code that builds a query. #2402  
So I don't want to do the work twice. To solve #2402  I will already use the new technique.

PaulHigin reviewed (commented):
We also need tests for this new functionality.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Diagnostics/GetEventCommand.cs:
```diff
@@ -1538,6 +1542,41 @@ private string HandleDataHashValue(Object value)
+            else
+            {
+                ret += string.Format(CultureInfo.InvariantCulture,
+                                         "((EventData[Data[@Name='{0}']='{1}']) or (UserData/*/{0}='{1}'))",
```
> "((EventData[Data[@Name='{0}']='{1}']) or (UserData/*/{0}='{1}'))"

Let's not repeat this twice. Maybe it's better to hold this format in a const variable.

iSazonov replied:
Fixed!

daxian-dbw replied:
Thanks!

daxian-dbw commented:
@iSazonov I have manually verified your code changes and it works as expected, thanks for fixing it! Now we are waiting on you to address the PR comments.
At the same time, @JamesWTruher is following up on the `.evtx` test file (thanks Jim!). Once we have that test file ready and merged, you can add test for this PR.

JamesWTruher commented:
I was able to construct a set of tests for this - it required an EVTX file - I'll create a PR for that (the test which uses filterhash is commented out - take a look, and if it's good, you can uncomment the test. I don't want to mark it as pending because on linux it means that tests would be marked both pending and skipped which is not allowed.

iSazonov committed:
Removed double cast and repeated string converted to const

1. Removed double cast of dataArray
2. Repeated string converted to const NamedDataTemplate

iSazonov committed:
Uncomment test for FilterHashtable

Test work good

iSazonov commented:
@JamesWTruher Great test! Thanks!

Required addition.
This test checks only the first part of my fixes.
>$filter = @{ path = "$eventLogFile"; Param2 = "Windows x64"}
The second part would look like this
>$filter = @{ path = "$eventLogFile"; Param2 = **"Windows x64","Windows x32"**}
as the named parameter by definition are String[].

iSazonov committed:
Removed ALL double cast of dataArray

daxian-dbw commented:
Thanks @iSazonov for the updates. The EVTX file is now available, please add another test to cover all your changes. Now we are waiting on you to address the `StringBuilder` comment and add the additional test.

msftclas commented:
Hi __@iSazonov__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

JamesWTruher commented:
Please feel free to add these tests in your PR, since the test code is now all merged, you should be able to add tests you think are good

---
In reply to: [250667902](https://github.com/PowerShell/PowerShell/pull/2373#issuecomment-250667902) [](ancestors = 250667902)

daxian-dbw commented:
@iSazonov thanks for your replies to the review comments. As for the additional test, please see @JamesWTruher 's comment.

iSazonov commented:
@JamesWTruher I can add second test, but it requires additional events in the evtx file. Now there are only three events, only one is suitable for new test and need at least two.
>$filter = @{ path = "$eventLogFile"; Param2 = "Windows x64","Windows x32"}

- here for the "Windows x64" there are a event and for "Windows x32" a event does not exists in evtx.

JamesWTruher commented:
I think you can use DriverName (it's in UserData.SetupInstallPrinterDriver) - it's not Param, but it should still work:
```
<Event xmlns='http://schemas.microsoft.com/win/2004/08/events/event'>
...
<UserData>
<SetupInstallPrinterDriver xmlns='http://manifests.microsoft.com/win/2005/08/windows/printing/spooler/core/events'>
...
<DriverName>Microsoft enhanced Point and Print compatibility driver</DriverName>
...
</SetupInstallPrinterDriver>
</UserData>
</Event>

<Event xmlns='http://schemas.microsoft.com/win/2004/08/events/event'>
...
<UserData>
<SetupInstallPrinterDriver ...
<DriverName>Remote Desktop Easy Print</DriverName>
... 
</SetupInstallPrinterDriver>
</UserData>
</Event>
```

daxian-dbw commented:
@iSazonov Please see @JamesWTruher's comment about using 'DriverName' to test passing a string array value like `$filter = @{ path = "$eventLogFile"; DriverName = "Remote Desktop Easy Print", "Microsoft enhanced Point and Print compatibility driver" }`.
And also, could you please rebase your branch? Your PR contains some merge commits.

iSazonov committed:
Add test for Get-WinEvent -FilterHashtable for named params

iSazonov commented:
1. Sorry for the delay the #2425 it stopped me. Now all test ready.
2. Sorry for the refuse I only learn git. Now I made rebase.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Diagnostics/Get-WinEvent.Tests.ps1:
```diff
@@ -95,7 +94,26 @@ Describe 'Get-WinEvent' -Tags "CI" {
+            # this relies on apriori knowledge about the log file
+            # the provided log file has been edited to remove MS PII, so we must use -ea silentlycontinue
+            $eventLogFile = [io.path]::Combine($PSScriptRoot, "assets", "Saved-Events.evtx")
+            $filter = @{ path = "$eventLogFile"; ; PackageAware="Not package aware"; DriverName = "Remote Desktop Easy Print", "Microsoft enhanced Point and Print compatibility driver" } 
```
> @{ path = "$eventLogFile"; ; PackageAware= ...

A small typo at "; ;". Can you please fix it?

iSazonov replied:
Fixed!

daxian-dbw commented:
@iSazonov Thanks for the update! It look mostly good except for one small typo (please see my comment above). Once you fix the typo, the PR will be ready to merge.

iSazonov committed:
Correct typo

iSazonov commented:
@daxian-dbw Small question about planned support of <Suppress> and use StringBuilder. In fact, we need to build a **XML** query: is it acceptable to use System.Xml.XmlWriter?

daxian-dbw reviewed (approved):
LGTM.

daxian-dbw commented:
@iSazonov Performance difference between `XmlWriter` and `StringBuilder` shouldn't be a consideration here, and they both will avoid generating a bunch of new strings like the current code does (Yay! :)). So you should pick the solution that you think is more maintainable.

`XmlWriter` make it easier to not generate any malformed XML queries, especially when the query is complex, but on the other hand it possibly will make the code less concise, for example, you may end up repeating following code snippet in many places. So if you decide to go with `XmlWriter`, please try avoid repeat as much as possible 😄 (for example, maybe put the code snippet below in a method that you can call whenever you need to generate a `<Query>` element)
```powershell
$xmlWriter.WriteStartElement("QueryList")
$xmlWriter.WriteStartElement("Query")
$XmlWriter.WriteAttributeString("Id", $Id.ToString())
$XmlWriter.WriteAttributeString("Path", $Path)

$xmlWriter.WriteStartElement("Select")
$XmlWriter.WriteAttributeString("Path", $Path)
```

daxian-dbw commented:
@iSazonov Thanks again for fixing this issue with Get-WinEvent!

iSazonov commented:
@daxian-dbw Thanks for help and advices! This is my first code after 20 years 😊
